File: src/services/studentService.ts
import { apiClient } from './api';
import { supabase } from './supabase';
import { Student } from '../types';

class StudentService {
  private tableName = 'students';

  async getAllStudents(): Promise<Student[]> {
    try {
      const { data, error } = await supabase.from(this.tableName).select('*');
      if (error) throw error;
      return data as Student[];
    } catch (error: any) {
      throw new Error(error.message || 'Failed to fetch students');
    }
  }

  async getStudentById(id: string): Promise<Student> {
    try {
      const { data, error } = await supabase.from(this.tableName).select('*').eq('id', id).single();
      if (error) throw error;
      if (!data) throw new Error('Student not found');
      return data as Student;
    } catch (error: any) {
      throw new Error(error.message || 'Failed to fetch student');
    }
  }

  async getStudentsByGrade(grade: string): Promise<Student[]> {
    try {
      const { data, error } = await supabase.from(this.tableName).select('*').eq('grade', grade);
      if (error) throw error;
      return data as Student[];
    } catch (error: any) {
      throw new Error(error.message || 'Failed to fetch students by grade');
    }
  }

  async createStudent(studentData: Partial<Student>): Promise<Student> {
    try {
      const { data, error } = await supabase.from(this.tableName).insert([studentData]).select().single();
      if (error) throw error;
      return data as Student;
    } catch (error: any) {
      throw new Error(error.message || 'Failed to create student');
    }
  }

  async updateStudent(id: string, updateData: Partial<Student>): Promise<Student> {
    try {
      const { data, error } = await supabase.from(this.tableName).update(updateData).eq('id', id).select().single();
      if (error) throw error;
      return data as Student;
    } catch (error: any) {
      throw new Error(error.message || 'Failed to update student');
    }
  }

  async deleteStudent(id: string): Promise<void> {
    try {
      const { error } = await supabase.from(this.tableName).delete().eq('id', id);
      if (error) throw error;
    } catch (error: any) {
      throw new Error(error.message || 'Failed to delete student');
    }
  }

  async searchStudents(searchTerm: string): Promise<Student[]> {
    try {
      // In a real app, you would use a search service like Algolia
      // This is a simple implementation using Firestore
      const allStudents = await this.getAllStudents();
      const lowerSearchTerm = searchTerm.toLowerCase();
      
      return allStudents.filter(student => 
        student.firstName.toLowerCase().includes(lowerSearchTerm) ||
        student.lastName.toLowerCase().includes(lowerSearchTerm) ||
        student.email.toLowerCase().includes(lowerSearchTerm) ||
        student.studentId.toLowerCase().includes(lowerSearchTerm)
      );
    } catch (error: any) {
      throw new Error(error.message || 'Failed to search students');
    }
  }
}

export const studentService = new StudentService();
export default studentService;

File: lib/services/student_service.dart
import 'package:supabase_flutter/supabase_flutter.dart';
import '../models/student.dart';

class StudentService {
  final SupabaseClient _supabase = Supabase.instance.client;

  // Get all students
  Future<List<Student>> getAllStudents() async {
    final response = await _supabase
        .from('students')
        .select()
        .order('name', ascending: true);
    
    return (response as List)
        .map((json) => Student.fromJson(json))
        .toList();
  }

  // Get student by ID
  Future<Student?> getStudentById(String id) async {
    final response = await _supabase
        .from('students')
        .select()
        .eq('id', id)
        .maybeSingle();
    
    return response != null ? Student.fromJson(response) : null;
  }

  // Create student
  Future<Student> createStudent(Student student) async {
    final response = await _supabase
        .from('students')
        .insert(student.toInsert())
        .select()
        .single();
    
    return Student.fromJson(response);
  }

  // Update student
  Future<Student> updateStudent(String id, Map<String, dynamic> updates) async {
    final response = await _supabase
        .from('students')
        .update(updates)
        .eq('id', id)
        .select()
        .single();
    
    return Student.fromJson(response);
  }

  // Delete student
  Future<void> deleteStudent(String id) async {
    await _supabase
        .from('students')
        .delete()
        .eq('id', id);
  }

  // Search students by name or email
  Future<List<Student>> searchStudents(String query) async {
    final response = await _supabase
        .from('students')
        .select()
        .or('name.ilike.%$query%,email.ilike.%$query%')
        .order('name', ascending: true);
    
    return (response as List)
        .map((json) => Student.fromJson(json))
        .toList();
  }

  // Get students by grade
  Future<List<Student>> getStudentsByGrade(String grade) async {
    final response = await _supabase
        .from('students')
        .select()
        .eq('grade', grade)
        .order('name', ascending: true);
    
    return (response as List)
        .map((json) => Student.fromJson(json))
        .toList();
  }
}
